Add Developer guide for further development. (#147)
sunpy/sunpy-soar: docs/dev_guide/query.rst

.. _sunpy-soar-dev-guide-query:

***************************************
Request methods ``sunpy-soar`` supports
***************************************

sunpy-soar currently supports two REQUEST methods: ``doQuery`` and ``doQueryFilteredByDistance``.

``doQuery``: This is the standard method used when no specific distance attribute is included in the search query.
It performs a general query based on the provided parameters, retrieving the data that matches the criteria specified.

``doQueryFilteredByDistance``: This method is employed when a distance parameter is included in the search query.

> Unlike ``doQuery``, this method filters the entire query based on the specified distance value.

Reviewer: In [line 54](https://github.com/sunpy/sunpy-soar/pull/147/files#diff-bd2f146f3b7a10c959bf01e70ed195c923d3ee69a4e4c98dc987ccb1890fe12fR54) it reads "filters the entire database"

The time attribute is not required when using ``doQueryFilteredByDistance``, as the time is internally calculated based on the provided distance.
The distance value is appended to the end of the query using ``&DISTANCE(dmin, dmax)``, where ``dmin`` and ``dmax`` are Astropy units of distance.

> These values must fall within the range of 0.28 AU to 1.0 AU; otherwise, the query will not return any results.

Reviewer: According to [this comment](https://github.com/sunpy/sunpy-soar/pull/137#issuecomment-2260560819) SOAR does indeed not behave as I expect.

> These values must fall within the range of 0.28 AU to 1.0 AU; otherwise, the query will not return any results.

Author: @nabobalis I think this was the only comment left to be addressed?

> These values must fall within the range of 0.28 AU to 1.0 AU; otherwise, the query will not return any results.

Reviewer 2: There are a range of "outdated" comments from Eric but they don't seem to be addressed in the text?

.. _sunpy-soar-dev-guide-query:

***************************************
Request methods ``sunpy-soar`` supports
***************************************

sunpy-soar currently supports two REQUEST methods: ``doQuery`` and ``doQueryFilteredByDistance``.

``doQuery``: This is the standard method used when no specific distance attribute is included in the search query.

> It performs a generalquery based on the provided parameters, retrieving the data that matches the criteria specified.

Reviewer 2: ```suggestion
It performs a general query based on the provided parameters, retrieving the data that matches the criteria specified.
```

> ``doQueryFilteredByDistance``: This method is employed when a distance parameter is included in the search query.
> Unlike doQuery, this method filters the entire query based on the specified distance value.
> The time attribute is not required when using doQueryFilteredByDistance,as the time is internally
> calculated based on the provided distance. The distance value is appended to the end of the query using &DISTANCE(dmin, dmax),
> where dmin and dmax are Astropy units of distance. These values must fall within the range of 0.28 AU to 1.0 AU;
> otherwise, the query will not return any results.

Reviewer 2: ```suggestion
``doQueryFilteredByDistance``: This method is employed when a distance parameter is included in the search query.
Unlike ``doQuery``, this method filters the entire query based on the specified distance value.
The time attribute is not required when using ``doQueryFilteredByDistance``, as the time is internally calculated based on the provided distance.
The distance value is appended to the end of the query using ``&DISTANCE(dmin, dmax)``, where ``dmin`` and ``dmax`` are Astropy units of distance.
These values must fall within the range of 0.28 AU to 1.0 AU; otherwise, the query will not return any results.
```
How is the time internally calculated?

Author: I don't exactly know this but Jonathan Cook mentioned it in #134 issue.

> Example:

Reviewer 2: ```suggestion
Using the example below,
```

.. code-block:: python

    import astropy.units as u
    import sunpy.net.attrs as a
    from sunpy.net import Fido

    instrument = a.Instrument("RPW")
    level = a.Level(2)
    distance = a.soar.Distance(0.28 * u.AU, 0.30 * u.AU)

>     result = Fido.search(instrument & level & distance)

Reviewer 2: We should check this result is output.

> Here the query's REQUEST type to "doQueryFilteredByDistance", which is a special method that filters the entire query
> based on the specified distance value. This method is used when a distance parameter is included in the search query,
> and the time attribute is not required.

Reviewer 2: ```suggestion
Here the query's "REQUEST" type to "doQueryFilteredByDistance", which is a special method that filters the entire query based on the specified distance value.
This method is used when a distance parameter is included in the search query, and the time attribute is not required.
```

> Example Query:

Reviewer 2: ```suggestion
The actual query this example produces is:
```

.. code-block:: python

>     Query="SELECT+*+FROM+v_sc_data_item+WHERE+instrument='RPW'+AND+level='L2'&DISTANCE(0.28,0.30)"

Reviewer 2: ```suggestion
   "SELECT+*+FROM+v_sc_data_item+WHERE+instrument='RPW'+AND+level='L2'&DISTANCE(0.28,0.30)"
```

> How can other Request methods be added?
> =======================================

Reviewer 2: ```suggestion
How can other request methods be added?
=======================================
```

To add support for additional request methods, you'll need to consider the impact they will have on the query structure.

> The REQUEST parameter in the query must be updated accordingly, and any necessary modifications to the query string should be
> made to accommodate the new method. If the new request method requires a specific attribute, you may need to add it as a class
> in the attrs.py file (if it doesn't already exist in `sunpy.net.attrs`). Additionally, a walker for this attribute
> will need to be created.

Reviewer 2: ```suggestion
The REQUEST parameter in the query must be updated accordingly, and any necessary modifications to the query string should be made to accommodate the new method.
If the new request method requires a specific attribute, you may need to add it as a class in the attrs.py file (if it doesn't already exist in `sunpy.net.attrs`).
Additionally, a walker for this attribute will need to be created.
```
You should link to the attrs file and provide an example of creating an attr walker for this.

           RPW rpw-tds-surv-hist2d    L2 2023-10-10 00:00:00.000 2023-10-11 00:00:00.000    3.382      None
           RPW        rpw-tnr-surv    L2 2023-10-10 00:00:00.000 2023-10-11 00:00:00.000  341.182      None
           RPW  rpw-lfr-surv-swf-e    L2 2023-10-10 00:00:00.000 2023-10-11 00:00:00.000  200.206      None
           RPW rpw-tds-surv-tswf-e    L2 2023-10-10 00:00:00.000 2023-10-11 00:00:00.000  173.184      None
    Length = 357 rows
    <BLANKLINE>
    <BLANKLINE>

Here the query's "REQUEST" type to "doQueryFilteredByDistance", which is a special method that filters the entire database based on the specified distance value.

The actual query this example produces is,

.. code-block:: python

   "SELECT+*+FROM+v_sc_data_item+WHERE+instrument='RPW'+AND+level='L2'&DISTANCE(0.28,0.30)"

How can other request methods be added?
=======================================

To add support for additional request methods, you'll need to consider the impact they will have on the query structure.
The REQUEST parameter in the query must be updated accordingly, and any necessary modifications to the query string should be made to accommodate the new method.

> If the new request method requires a specific attribute, you may need to add it as a class in the attrs.py file (if it doesn't already exist in `sunpy.net.attrs`).

Reviewer 2: I would link to the attrs.py file on main using a url

Reviewer 2: Or maybe just a filepath from the root of the git repo.
